### PR TITLE
[MM-47489] Move Shared Channels (Experimental) to Professional

### PR DESCRIPTION
**Pull request by mylonsuren**
#### Summary

Moves Shared Channels to professional license.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-47489

#### Release Note

```release-note
Moved Shared Channels (Experimental) to Professional
```

### CODE REVIEW

File: config/client.go [hunk 1]
```diff
@@ -199,12 +199,11 @@ func GenerateClientConfig(c *model.Config, telemetryID string, license *model.Li
 			props["CWSURL"] = *c.CloudSettings.CWSURL
 		}
 
-		if *license.Features.SharedChannels {
+		if license.SkuShortName == model.LicenseShortSkuProfessional || license.SkuShortName == model.LicenseShortSkuEnterprise {
+			*license.Features.SharedChannels = true
```
Comment (vish9812) [inline config/client.go:203]: It seems, instead of assignment, it should be part of the `if` condition check.

**Review by vish9812 — COMMENTED** [review]
Overall, I'm not sure about this. Since this feature is moving from enterprise to professional, I was expecting to be some checks for enterprise to be replaced by professional license checks. So, do you know how this feature was working just for the enterprise before?

File: config/client.go (hunk 1 above)
Comment (mylonsuren) [reply config/client.go:203]: Yeah that was my initial thought as well, but was unable to find how this was set up. I'm currently in the process of investigating this and will update this PR once it's ready again.

**Comment by ashishbhate** [comment]
>Overall, I'm not sure about this. Since this feature is moving from enterprise to professional, I was expecting to be some checks for enterprise to be replaced by professional license checks. So, do you know how this feature was working just for the enterprise before?


IIUC, A SKU (Professional, Enterprise etc.) is a collection of features (clustering, elastic search, message export etc.). So we can either check that the license has the feature or the license has the correct SKU. 

@neallred Is this correct? Asking you because I think this comes under Growth/Self-Serve's area of expertise.

**Comment by neallred** [comment]
@ashishbhate good call that we want features to be driven by license features rather than license sku name. The code that determines that is in customer-web-server, here: https://github.com/mattermost/customer-web-server/blob/master/licenses/licenses.go#L192. I made a PR to update it here: https://github.com/mattermost/customer-web-server/pull/1039

@marianunez , in the past when features have moved from a higher tier to a lower tier, do you know how we have handled *existing* licenses that won't have the updated access? Do add checks like the ones in this PR to retroactively grant them access, or do we just let it wait until renewal time for them to get the updated access?

**Comment by marianunez** [comment]
> @ashishbhate good call that we want features to be driven by license features rather than license sku name.

Actually we are moving away from having specific feature bits in the license and relying on the sku checks directly to reduce maintenance in updating CWS for any feature change and to be backward compatible with licenses that are out in the field and don't have this bit.

**Comment by marianunez** [comment]
> @marianunez , in the past when features have moved from a higher tier to a lower tier, do you know how we have handled _existing_ licenses that won't have the updated access? Do add checks like the ones in this PR to retroactively grant them access, or do we just let it wait until renewal time for them to get the updated access?

With checking the Sku, it is backward compatible with any existing licenses and they don't need to change the license but just upgrade to have access to the new features. 

For reference https://docs.google.com/document/d/1PhlIiGXbAl-sTRFXLRT5R71eSzUTA0HXAVNXPu1aAVU/edit#heading=h.281rlz9jxg8q

**Comment by neallred** [comment]
Sounds good! Closed that customer-web-server PR. For this PR then, I think we may still have legacy self hosted professional plans with valid licenses. So I think we should add `model.LicenseShortSkuE10` to the list of checks.

**Comment by mylonsuren** [comment]
> I think we may still have legacy self hosted professional plans with valid licenses. So I think we should add `model.LicenseShortSkuE10` to the list of checks.

@neallred Would we also need to add `model.LicenseShortSkuE20` or is that not required for legacy enterprise customers?

**Comment by neallred** [comment]
It is a good idea for purposes of being complete. I like this example: https://github.com/mattermost/mattermost-server/blob/master/model/license.go#L308-L313. I'm imagining we would do something similar here, like

```go
func (l *License) HasSharedChannels() bool {
	return l != nil &&
		(l.Features != nil && l.Features.SharedChannels != nil && *l.Features.SharedChannels) ||
		l.SkuShortName == LicenseShortSkuE10 ||
		l.SkuShortName == LicenseShortSkuE20 ||
		l.SkuShortName == LicenseShortSkuProfessional ||
		l.SkuShortName == LicenseShortSkuEnterprise
}
```

Then this function would be safe, complete, and reusable, even in the case that license or license.features is nil.

**Comment by marianunez** [comment]
> @neallred Would we also need to add `model.LicenseShortSkuE20` or is that not required for legacy enterprise customers?

@wiersgallak @esethna Is the expectation that shared channels is also available for E10 and E20 or is this intended to be for only new Professional licenses?

**Comment by kwiersgalla** [comment]
> Is the expectation that shared channels is also available for E10 and E20 or is this intended to be for only new Professional licenses?

@marianunez @mylonsuren This change will only go in the new packages and not backported to E10/E20. Thanks for confirming!

**Review by neallred — COMMENTED** [review]
I think we will still need to check `license.Features.SharedChannels` for the time being so that existing E20 licenses, which would have had this feature enabled, do not lose access to the feature.

Optionally, I still think we could define a helper like

```go
func (l *License) HasSharedChannels() bool {
	if  (l == nil) {
		return false
	}
	return (l.Features != nil && l.Features.SharedChannels != nil && *l.Features.SharedChannels) ||
		l.SkuShortName == LicenseShortSkuProfessional ||
		l.SkuShortName == LicenseShortSkuEnterprise
}
```

In addition to covering existing E20 licenses with this feature, it would make other code a little more DRY.

**Review by ashishbhate — APPROVED** [review]

**Comment by mylonsuren** [comment]
@vish9812 @neallred Should be good for another review here 😄

**Review by neallred — APPROVED** [review]
LGTM, thanks @mylonsuren !

**Review by vish9812 — APPROVED** [review]

**Pull request merged**